### DISCUSSION TITLE
Speed test: How's it work?

### DISCUSSION THREAD

**Discussion by damccull** [ravendb/ravendb#15196] [Q&A] [answered]
I'm looking through the code for the speed test and as I understand it (hints from the Cluster Speed Test documentation), it behaves like this:

* Speed test only happens if mode is FastestNode.
* Speed test is triggered by new topology, entry into FastestNode mode, or a one-minute timer as long as FastestNode is active.
* Speed test is performed in the first two (non-timer) cases by sending the user supplied "READ" request to all servers and marking the time it takes them each to complete the query, then marking the fastest.

My two questions are:

1) How is the speed test completed in the one-minute timer situation? Does the RequestExecutor or NodeSelector cache the most recent "READ" command and keep re-sending it every minute? Or does it use some pre-defined query I've not yet found?
2) Am I correctly understanding this procedure?

**Comment by ayende** [comment, marked as answer]
The relevant code is here: 

https://github.com/ravendb/ravendb/blob/8f1df2b4d64aa6be6c41015e78dc36838c8c1f56/src/Raven.Client/Http/RequestExecutor.cs#L1060-L1063

The way it work is that we have a flag (`InSpeedTestPhase`) that is raised once a minute by a timer.

The once a minute is here:
https://github.com/ravendb/ravendb/blob/b6af2d6fd65fd96cb96aa53b9e9399bc48b44e96/src/Raven.Client/Http/NodeSelector.cs#L262

When that flag is raised, if we have a read request (user provided, not predefined) we'll send it to all the nodes here:
https://github.com/ravendb/ravendb/blob/8f1df2b4d64aa6be6c41015e78dc36838c8c1f56/src/Raven.Client/Http/RequestExecutor.cs#L1060-L1063

The first to respond is handled here:
https://github.com/ravendb/ravendb/blob/8f1df2b4d64aa6be6c41015e78dc36838c8c1f56/src/Raven.Client/Http/RequestExecutor.cs#L1301

Where we setup the next check in a minute.

Once we selected the fastest node, we remember that here: 
https://github.com/ravendb/ravendb/blob/b6af2d6fd65fd96cb96aa53b9e9399bc48b44e96/src/Raven.Client/Http/NodeSelector.cs#L17-L18

And it is returned on `GetFastestNode`

**Reply by damccull** [reply]
That seems simple enough. So you're not always sending a command every minute, just raising the rest flag each minute and then the next read command will do the test? Thanks for the reply.

**Reply by ayende** [reply]
Yes, that is correct.
